allspice/utils/list_components.py

```diff
@@ -368,57 +453,111 @@ def _resolve_prjpcb_relative_path(schdoc_path: str, prjpcb_path: str) -> str:
 
 
 def _build_schdoc_hierarchy(
-    sheets_to_refs: dict[str, list[dict]],
+    schematic_document_jsons: dict[str, dict],
+    device_sheet_jsons: dict[str, dict],
 ) -> tuple[set[str], SchdocHierarchy]:
     """
     Build a hierarchy of sheets from a mapping of sheet names to the references
     of their children.
 
-    The output of this function is a tuple of two values:
+    :param document_jsons: A mapping of document sheet paths from project root
+        to their JSON.
+    :param device_sheet_jsons: A mapping of device sheet *names* to their JSON.
+    :returns: The output of this function is a tuple of two values:
 
     1. A set of "independent" sheets, which can be taken to be roots of the
-    hierarchy.
-
-    2. A mapping of each sheet that has children to a list of tuples, where each
-    tuple is a child sheet and the number of repetitions of that child sheet in
-    the parent sheet. If a sheet has no children and is not a child of any other
-    sheet, it will be mapped to an empty list.
+       hierarchy.
+    2. A mapping of each sheet that has children to a list of tuples, where
+       each tuple is a child sheet and the number of repetitions of that child
+       sheet in the parent sheet. If a sheet has no children and is not a child
+       of any other sheet, it will be mapped to an empty list.
     """
 
     hierarchy = {}
 
-    # We start by assuming all sheets are independent.
-    independent_sheets = set(sheets_to_refs.keys())
+    schematic_document_entries = {
+        schdoc_file: [value for value in schdoc_json.values() if isinstance(value, dict)]
+        for schdoc_file, schdoc_json in schematic_document_jsons.items()
+    }
+    schematic_document_refs = {
+        schdoc_file: [entry for entry in entries if entry.get("type") == "SheetRef"]
+        for schdoc_file, entries in schematic_document_entries.items()
+    }
+
+    device_sheet_entries = {
+        device_sheet_name: [
+            value for value in device_sheet_json.values() if isinstance(value, dict)
+        ]
+        for device_sheet_name, device_sheet_json in device_sheet_jsons.items()
+    }
+    device_sheet_refs = {
```

Comment:
Can you explain how device sheet refs in device sheets work for me? For example will all device sheets referenced be in the project file?

Reply:
Yes, all Device Sheets are listed in the project file, so we can pull the list from the project file, find the file either in the repo or in a design reuse repository, and map the list of device sheet names to their JSONs. Then we can treat them *mostly* as regular documents, with two major exceptions: 

- Device sheets can only reference other device sheets, since they are external to the project.
- Device sheets are referenced by their name in the SheetRef's `filename` field, while Documents are referenced by their relative path from the file.